### PR TITLE
Optimize histogram iterators

### PR DESCRIPTION
**Pull request by fpetkovski**
Histogram iterators allocate new objects in the AtHistogram and AtFloatHistogram methods, which makes calculating rates over long ranges expensive.

In #13215 we allowed an existing object to be reused when converting an integer histogram to a float histogram. This commit follows the same idea and allows injecting an existing object in the `AtHistogram` and `AtFloatHistogram` methods. When the injected value is nil, iterators allocate new histograms, otherwise they populate and return the injected object.

The commit also adds a CopyFrom method to `Histogram` and `FloatHistogram` which is used in the `BufferedSeriesIterator` to overwrite items in the ring instead of making new copies.

With these changes, memory and CPU for histogram queries look as shown bellow. I also took measurement from the main branch prior to #13215.

```
name \ time/op                                        pre-13215.out  main.out     new.out
NativeHistograms/sum-8                                   477ms ± 1%   484ms ± 2%   496ms ± 2%
NativeHistograms/sum_rate_with_short_rate_interval-8     1.02s ± 1%   0.91s ± 6%   0.60s ± 0%
NativeHistograms/sum_rate_with_long_rate_interval-8      1.43s ± 1%   1.31s ± 3%   1.01s ± 1%

name \ alloc/op                                       pre-13215.out  main.out     new.out
NativeHistograms/sum-8                                   414MB ± 0%   414MB ± 0%   428MB ± 0%
NativeHistograms/sum_rate_with_short_rate_interval-8    1.17GB ± 0%  0.85GB ± 0%  0.23GB ± 0%
NativeHistograms/sum_rate_with_long_rate_interval-8     1.27GB ± 0%  0.95GB ± 0%  0.33GB ± 0%

name \ allocs/op                                      pre-13215.out  main.out     new.out
NativeHistograms/sum-8                                   5.16M ± 0%   5.16M ± 0%   5.18M ± 0%
NativeHistograms/sum_rate_with_short_rate_interval-8     21.2M ± 0%   14.7M ± 0%    4.6M ± 0%
NativeHistograms/sum_rate_with_long_rate_interval-8      22.8M ± 0%   16.2M ± 0%    6.1M ± 0%
```

### CODE REVIEW

File: model/histogram/float_histogram.go [hunk 1]
```diff
@@ -53,28 +53,57 @@ type FloatHistogram struct {
...
+	if len(h.NegativeBuckets) != 0 {
 		c.NegativeBuckets = make([]float64, len(h.NegativeBuckets))
 		copy(c.NegativeBuckets, h.NegativeBuckets)
 	}
 
 	return &c
 }
 
+// CopyFrom makes a deep copy of the given FloatHistogram into the receiver object.
+func (h *FloatHistogram) CopyFrom(from FloatHistogram) {
```
Comment (bwplotka) [inline model/histogram/float_histogram.go:86]: Hm, depending on what's more consistent/easier we could have only "two" copy methods:

`func (h *FloatHistogram) Copy(buf *FloatHistogram) *FloatHistogram {`
and 
`func (h *FloatHistogram) CopyToSchema(targetSchema int32, buf *FloatHistogram) *FloatHistogram {`

I feel this is what we do in other places in Prometheus e.g. ChunkIterators. 

Additionally I am not sure about extra value copy here in "from FloatHistogram". We already have it on heap so why copying things to (potentially to stack) for this function? Probably no perf difference, just weird? I think if we keep CopyFrom it should be `CopyFrom(from *FloatHistogram`) perhaps 🤔

**Review by bwplotka — COMMENTED** [review]
Nice! One question around readability/consistency, I would propose a different "Copy" function change here 🤔 

Otherwise LGTM!

**Comment by beorn7** [comment]
Great stuff @fpetkovski, us usual. And thanks @bwplotka for starting the review.

I'm still on vacation for another week and won't have time for an in-depth review.

@bwplotka @krajorama feel free to go ahead with the review as you see fit.

File: model/histogram/float_histogram.go (hunk 1 above)
Comment (bwplotka) [reply model/histogram/float_histogram.go:86]: What about "buf" idea and changing existing methods?

Comment (fpetkovski) [reply model/histogram/float_histogram.go:86]: I feel like the "buf" API could indicate that the passed in histogram could be safely reused after calling Copy, leading to subtle data corruption bugs. Separate `CopyTo` and `CopyToSchema` methods do make sense though, but I don't really have a strong opinion. I see that we already have a `func (ls *Labels) CopyFrom(b Labels)` method though, so not sure what's more consistent with the rest of the codebase.

Comment (krajorama) [reply model/histogram/float_histogram.go:86]: I tried to think through how merging CopyFrom (or more likely CopyTo) with CopyToSchema work. In case the requested schema is equal to the current schema, it's simply an optimization like in AtHistogram. In case the schema is different (lower), then I'd expect to reuse the spans/buckets in the target histogram, which you could do by first copying over the target histogram then executing `reduceResolution` with `inplace==true`. 

So technically it seems possible, but I'm not sure we want to complicate the code and testing as the use cases don't intersect at the moment. You'd want to use CopyTo to fill the samples being input to the engine and then CopyToSchema is used when calculating e.g. rate() in the engine, and there it's unlikely you want reuse.

To summarize: agree with Filip on keeping separate and making simpler `CopyTo`, want to pass pointers everywhere for consistency.   

When looking into this, I was reminded that histogram Add/Sub can change samples in place by altering the schema, which makes reusing those series in other parts of the query execution less precise. Basically we're assuming that we'll never optimize away loading series for each sub query individually :)

File: tsdb/chunkenc/float_histogram.go [hunk 2]
```diff
@@ -728,27 +728,50 @@ func (it *floatHistogramIterator) At() (int64, float64) {
...
-		ZeroCount:        it.zCnt.value,
-		Sum:              it.sum.value,
-		ZeroThreshold:    it.zThreshold,
-		Schema:           it.schema,
-		PositiveSpans:    it.pSpans,
-		NegativeSpans:    it.nSpans,
-		PositiveBuckets:  it.pBuckets,
-		NegativeBuckets:  it.nBuckets,
+	if fh == nil {
+		it.atFloatHistogramCalled = true
```
Comment (krajorama) [inline tsdb/chunkenc/float_histogram.go:740]: nice , lgtm to only set it here and not the following branch

File: storage/buffer.go [hunk 3]
```diff
@@ -378,7 +381,11 @@ func (it *SampleRingIterator) AtFloatHistogram(fh *histogram.FloatHistogram) (in
 	if it.fh == nil {
 		return it.t, it.h.ToFloat(fh)
 	}
-	return it.t, it.fh
+	if fh != nil {
+		fh.CopyFrom(*it.fh)
+		return it.t, fh
+	}
+	return it.t, it.fh.Copy()
```
Comment (krajorama) [inline storage/buffer.go:388]: This copy is new, is it there for consistency or was there a bug?

File: storage/buffer.go [hunk 4]
```diff
@@ -672,7 +679,12 @@ func addH(s hSample, buf []hSample, r *sampleRing) []hSample {
 		}
 	}
 
-	buf[r.i] = s
+	buf[r.i].t = s.t
+	if buf[r.i].h == nil {
+		buf[r.i].h = s.h.Copy()
```
Comment (krajorama) [inline storage/buffer.go:684]: Same question as above, these copies are new, do we need them?

File: storage/buffer.go [hunk 5]
```diff
@@ -711,7 +723,12 @@ func addFH(s fhSample, buf []fhSample, r *sampleRing) []fhSample {
 		}
 	}
 
-	buf[r.i] = s
+	buf[r.i].t = s.t
+	if buf[r.i].fh == nil {
+		buf[r.i].fh = s.fh.Copy()
```
Comment (krajorama) [inline storage/buffer.go:728]: Same question as above, these copies are new, do we need them?

File: storage/buffer.go (hunk 4 above)
Comment (fpetkovski) [reply storage/buffer.go:684]: This copy is needed because we now use a single histogram object for decoding chunks (the `hReader` and `fhReader` variables), so we have to copy those samples when we store them in the buffer. Otherwise, the next read will overwrite buffered samples.

File: storage/buffer.go (hunk 3 above)
Comment (fpetkovski) [reply storage/buffer.go:388]: This copy is not necessary here, but I thought it is safer to do it than not to. When we create a new histogram slice with `getHPointSlice`, the items in it will be `nil` pointers. If we pass in `nil` as the argument, we will get a reference to a sample in the buffer, and it will be easy to overwrite it by mistake if we use the sample as an argument to `floatHistogramIterator.AtFloatHistogram`. 

The alternative would be to never use `nil` as the argument and allocate all histogram objects at the call site (in `matrixIterSlice`) so that we trigger the `CopyFrom` branch. I would be fine with either solution, this was just the simplest one to implement.

File: model/histogram/float_histogram.go (hunk 1 above)
Comment (fpetkovski) [reply model/histogram/float_histogram.go:86]: Sounds good, I will replace `CopyFrom` with `CopyTo` for consistency with other histogram methods.

Comment (fpetkovski) [reply model/histogram/float_histogram.go]: I've replaced CopyFrom with CopyTo, let me know if this looks good now.

File: storage/buffer.go (hunk 4 above)
Comment (krajorama) [reply storage/buffer.go:684]: I see, does that mean we're copying twice? Once in hReader from the chunk, then copy again into the sample buffer?

(probably a next step to optimize away if yes)

File: storage/buffer.go (hunk 3 above)
Comment (krajorama) [reply storage/buffer.go:388]: I have the feeling that this takes away some optimization possibilities to delay the copy (or avoid copy if the caller uses the data as read only) and is overly defensive. On the other hand it's for sure safer. So I guess as long as performance is not hit, it's fine.

Comment (fpetkovski) [reply storage/buffer.go:388]: I agree, it's definitely suboptimal and something we can look into next. However, I did some local testing with the new branch and the copy here was needed unless I passed in an object that could be filled. Right now histograms tests are not comprehensive enough to catch regressions of this sort. 

I think the better solution is for `getHPointSlice` to return a slice of value objects rather than pointers, so that when we get a slice it will always have non-nil values.

Comment (krajorama) [reply storage/buffer.go:388]: ack

File: storage/buffer.go (hunk 4 above)
Comment (krajorama) [reply storage/buffer.go:684]: I see, took me a while to understand how hReader is used, as it's not directly used to read from

**Review by krajorama — APPROVED** [review]
LGTM.

I think there's more opportunity to optimize as we discussed. In general  from the PromQL engine point of view we always ask for float histograms, so in theory we should be able to pass that request down to the chunkenc.AtFloatHistogram call that can read the data from a floathistogram chunk or integer histogram chunk directly. But if I understand correctly, right now we'd copy to an integer histogram in a buffer first, then convert in a new step.

**Review by beorn7 — APPROVED** [review]
Thank you. Will merge once @krajorama 's final nit has been addressed.

**Comment by fpetkovski** [comment]
Let me take a look at the test failure.

**Comment by beorn7** [comment]
Yeah, the test failure looks like a real one. (It's via the testing framework, not a conventional unit test. I plan to add way more PromQL tests for native histograms in this way. It's easy.)

**Review by beorn7 — APPROVED** [review]
Thank you very much. I think everything should be addressed now. Will merge.
